db/src/query.ts

```diff
@@ -0,0 +1,441 @@
+    if (typeof pattern === 'string') {
+      return {
+        [filterSymbol]: true,
+        operator: 'matches',
+        path: this.parts,
+        pattern,
+        caseInsensitive,
+      };
+    } else if (pattern instanceof RegExp) {
+      return {
+        [filterSymbol]: true,
+        operator: 'matches',
+        path: this.parts,
+        pattern: pattern.source,
+        caseInsensitive: pattern.flags.includes('i'),
+      };
+    }
+    throw new TypeError('Expected pattern to be a RegExp or string');
+  }
+
+  /**
+   * String starts with prefix
+   * @return - a filter
+   */
+  public startsWith(prefix: string): StartsWithFilter {
+    return {
+      [filterSymbol]: true,
+      operator: 'startsWith',
+      path: this.parts,
+      value: prefix,
+    };
+  }
+
+  /**
+   * Casts to a specific doc type. Use for typed paths with typescript.
+   * @return - a filter
+   */
+  public as<T>(): Doc<T> {
+    return Path.proxied(this.parts) as any;
+  }
+}
+
+type PathProxy = Path & Record<Key, Path>;
+
+interface CastablePath {
+  as<T>(): Doc<T>;
+}
+
+interface EquatablePath<T extends Equatable> extends CastablePath {
+  eq(x: T): EqFilter;
+  ne(x: T): NeFilter;
+}
+
+interface ComparablePath<T extends Comparable> extends EquatablePath<T> {
+  gt(x: T): GtFilter;
+  gte(x: T): GteFilter;
+  lt(x: T): LtFilter;
+  lte(x: T): LteFilter;
+}
+
+interface StringPath extends ComparablePath<string> {
+  matches(pattern: string, caseInsensitive?: boolean): MatchesFilter;
+  matches(pattern: RegExp): MatchesFilter;
+  startsWith(prefix: string): StartsWithFilter;
+}
+
+type NumberPath = ComparablePath<number>;
+type DatePath = ComparablePath<Date>;
+
+type BooleanPath = EquatablePath<boolean>;
+
+interface NullPath extends CastablePath {
+  isNull(): IsNullFilter;
+}
+
+interface MaybePath extends CastablePath {
+  exists(): Filter;
+}
+
+type Doc<T> = T extends Record<string, unknown> ? Required<{
+  // technically type should have field(P) => Doc<T[P]> but that's not supported in typescript
+  [P in keyof T]: Doc<T[P]> & MaybePath;
+}>
+  : T extends Array<infer U> ? { [idx: number]: Doc<U> & MaybePath }
+  : T extends null ? NullPath
+  : T extends number ? NumberPath
+  : T extends string ? StringPath
+  : T extends boolean ? BooleanPath
+  : T extends Date ? DatePath
+  : never;
+
+export function typedValue<T>() {
+  return Path.proxied(['value']) as unknown as Doc<T>;
+}
+
+export const key = Path.proxied(['key']) as unknown as Doc<string>;
+export const value = Path.proxied(['value']);
+
+type NonEmptyArray<T> = [T, ...T[]];
+
+function checkFilters(...filters: any[]) {
+  if (filters.length === 0) {
+    throw new IllegalArgumentError('Expected at least 1 filter');
+  }
+  for (const f of filters) {
+    if (!f[filterSymbol]) {
+      throw new TypeError('Given filter is invalid');
+    }
+  }
+}
+
+export function all(...filters: NonEmptyArray<Filter>): AndFilter {
+  checkFilters(...filters);
```

Comment:
After our F2F we agreed that it would be very hard for a user to pass a filter created not with one of the exposed functions.
In the future the filters will be serialized and sent over network which will make it impossible to verify filters were created using one of the provided constructors.

Reply:
Agreed.  It is not possible to write (reasonable) code using the API that allows an injection using punning (passing an object where a string or number was expected), as values only go in leaves.